src/py/run_generator.rs

```diff
@@ -0,0 +1,290 @@
+            ErrorCode::InvalidPubkey => 10,
+            ErrorCode::InvalidMessage => 10,
+            ErrorCode::InvalidParentId => 10,
+            ErrorCode::InvalidConditionOpcode => 10,
+            ErrorCode::InvalidCoinAnnouncement => 10,
+            ErrorCode::InvalidPuzzleAnnouncement => 10,
+            ErrorCode::InvalidCondition => 10,
+            ErrorCode::InvalidCoinAmount => 10,
+            ErrorCode::AssertHeightAbsolute => 14,
+            ErrorCode::AssertHeightRelative => 13,
+            ErrorCode::AssertSecondsAbsolute => 15,
+            ErrorCode::AssertSecondsRelative => 105,
+            ErrorCode::AssertMyAmountFailed => 116,
+            ErrorCode::AssertMyPuzzlehashFailed => 115,
+            ErrorCode::AssertMyParentIdFailed => 114,
+            ErrorCode::AssertMyCoinIdFailed => 11,
+            ErrorCode::AssertPuzzleAnnouncementFailed => 12,
+            ErrorCode::AssertCoinAnnouncementFailed => 12,
+            ErrorCode::ReserveFeeConditionFailed => 48,
+            ErrorCode::DuplicateOutput => 4,
+            ErrorCode::DoubleSpend => 5,
+            ErrorCode::CostExceeded => 23,
+        };
+        ret.to_object(py)
+    }
+}
+
+// returns the cost of running the CLVM program along with the list of NPCs for
+// the generator/spend bundle. Each SpendConditionSummary is a coin spend along with its
+// conditions
+#[allow(clippy::too_many_arguments)]
+#[pyfunction]
+pub fn run_generator(
+    py: Python,
+    program: &[u8],
+    args: &[u8],
+    quote_kw: u8,
+    apply_kw: u8,
+    opcode_lookup_by_name: HashMap<String, Vec<u8>>,
+    max_cost: Cost,
+    flags: u32,
+) -> PyResult<(Option<ErrorCode>, Vec<PyNPC>, Cost)> {
+    let mut allocator = Allocator::new();
+    let f_lookup = f_lookup_for_hashmap(opcode_lookup_by_name);
+    let strict: bool = (flags & STRICT_MODE) != 0;
+    let f = OperatorHandlerWithMode::new(f_lookup, strict);
+    let program = node_from_bytes(&mut allocator, program)?;
+    let args = node_from_bytes(&mut allocator, args)?;
+
+    let r = py.allow_threads(
+        || -> Result<(Option<ErrorCode>, Cost, Vec<SpendConditionSummary>), EvalErr> {
+            let Reduction(cost, node) = run_program(
+                &mut allocator,
+                program,
+                args,
+                &[quote_kw],
+                &[apply_kw],
+                max_cost,
+                &f,
+                None,
+            )?;
+            // we pass in what's left of max_cost here, to fail early in case the
+            // cost of a condition brings us over the cost limit
+            match parse_spends(&allocator, node, max_cost - cost, flags) {
+                Err(ValidationErr(_, c)) => {
+                    Ok((Some(c), 0_u64, Vec::<SpendConditionSummary>::new()))
+                }
+                Ok(spend_list) => Ok((None, cost, spend_list)),
+            }
+        },
+    );
+
+    let mut ret = Vec::<PyNPC>::new();
+    match r {
```

Comment:
You could move all this stuff up to the threaded area since you only use `py` in one of the error cases (and in that case, you could reacquire the `py` lock without much concern, since it should happen only rarely).

Reply:
right. My original intention was to use the `Py` object to create native python objects, but I couldn't get that to work